crates/env/src/backend.rs

```diff
@@ -115,8 +115,15 @@ pub trait EnvBackend {
     where
         R: scale::Encode;
 
-    /// Prints the given contents to the console log.
-    fn println(&mut self, content: &str);
+    /// Emit a custom debug message.
+    ///
+    /// The message is appended to the debug buffer which is then supplied to the calling RPC
+    /// client. This buffer is also printed as a debug message to the node console if the
+    /// `debug` log level is enabled for the `runtime::contracts` target.
+    ///
+    /// This is a no-op if debug message recording is disabled which is always the case
+    /// when the code is executing on-chain.
```

Comment:
This statement is only true from the viewpoint of pallet-contracts. Contracts will pay for the overhead of calling into the supervisor when a debug message will be emitted during on-chain execution (it will do nothing but call into the supervisor and return). 

I recommend integrating one mitigation into this PR: Evaluate the return value of the `seal_debug_message` and if it is `false` prevent all further calls of this function.

Reply:
I think we should handle this on a compile time level. So only if we compile using some crate feature (which is enabled by default) the debug message will actually do something useful. Otherwise it will be a no-op. The `cargo-contract` tool should then disable the `debug` crate feature by default upon doing a release build and should leave it when building a debug build for the contract.

Reply:
This makes sense. I guess you will want different log levels then for message you still want to be in the release build (to help users with wrong inputs).

That said, you can have both. I think for messages that should indeed be left inside a production contract to help real users during production use you will need this runtime check in order to prevent weight costs during on-chain execution.

Reply:
> I think we should handle this on a compile time level.

I've done the runtime check but might leave the compile time check for a follow up since it would require some changes to `cargo-contract` which builds by `--release` by default.